### PR TITLE
Pluggable signer API [WIP - IdCard+MobileID]

### PR DESCRIPTION
**Pull request by ykshatroff**
This is a new API proposal for signing. It is WIP so may not be perfect in every aspect, but any ideas are highly welcome.

Start here: https://github.com/thorgate/django-esteid/tree/pluggable-signer/esteid/signing

Note:
The test views for the new signers are located at <yourhost:yourport>/new/ and ATM only ID Card and Mobile ID are hooked.

### CODE REVIEW

File: esteid/signing/views.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    def get(self, request: "RequestType", *args, **kwargs):
+        """
+        Handles finalize requests that do not require input data
+        """
+        return self.finish_session(request, *args, **kwargs)
+
+    @staticmethod
+    def parse_request(request):
+        """
+        Parses PATCH/POST request bodies as JSON or urlencoded, and assigns `request.data`.
```
Comment (Jyrno42) [inline esteid/signing/views.py:183]:
```suggestion
        Parses PATCH/POST request bodies as JSON or urlencoded and returns it.
```

File: esteid/flowtest/signer.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        return container.finalize()
+
+
+class MyPostSigner(MySigner):
+    """
+    Requires POST method parameters to init and finalize
+    """
+
+    def setup(self, initial_data: dict = None):
+        print(initial_data)
```
Comment (TonisPiip) [inline esteid/flowtest/signer.py:29]: Debug print trying to sneak though?

Comment (TonisPiip) [inline esteid/flowtest/signer.py]: Debug print trying to sneak though?

Comment (TonisPiip) [inline esteid/flowtest/signer.py]: Debug print trying to sneak through?

File: esteid/signing/tests/test_types.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        ),
+        pytest.param(
+            {"digest_b64": "", "temp_signature_file": "", "temp_container_file": "", "timestamp": 0},
+            None,
+            id="empty values",
+        ),
+        pytest.param(
+            {"digest_b64": "dGVzdA==", "temp_signature_file": "asdf", "temp_container_file": "asdf", "timestamp": 0},
+            None,
+            id="empty values",
```
Comment (TonisPiip) [inline esteid/signing/tests/test_types.py]: wrong param ID?

File: esteid/flowtest/signer.py (hunk 2 above)
Comment (ykshatroff) [reply esteid/flowtest/signer.py]: Thanks, fixed :)

File: esteid/signing/tests/test_types.py (hunk 3 above)
Comment (ykshatroff) [reply esteid/signing/tests/test_types.py]: True, renamed :)

File: esteid/urls.py [hunk 4]
```diff
@@ -19,16 +21,19 @@
 
 urlpatterns = []
 
-if settings.DEBUG:
+if settings.ROOT_URLCONF == __name__:
```
Comment (TonisPiip) [inline esteid/urls.py]: this looks odd.

File: esteid/templates/esteid/test-new.html [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+            The signatures will not be valid unless the test certificates are used to validate them.
+        </li>
+        <li>
+            ID Card signing can be done with real ID cards, but such a signature can not be validated.
+        </li>
+        <li>
+            ID Card signing requires the chrome-token-signing for chrome or firefox.
+        </li>
+        <li>
+            ID Card signing requires HTTPS. Please refer to the README.
```
Comment (iharthi) [inline esteid/templates/esteid/test-new.html:33]: Just a thought :thinking: - maybe we can reuse the part of thorgate-test-assignment here and display readme directly here instead of referring to readme. Probably a bad idea, since we use some library there to render markdown...

File: esteid/templates/esteid/test-new.html [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+
+    <ul>
+        <li>
+            Mobile ID signing is done with Demo Mobile ID service, you can only use the
+            <a href="https://github.com/SK-EID/MID/wiki/Test-number-for-automated-testing-in-DEMO" target="_blank">demo
+                test numbers</a>.
+            The signatures will not be valid unless the test certificates are used to validate them.
+        </li>
+        <li>
+            ID Card signing can be done with real ID cards, but such a signature can not be validated.
```
Comment (iharthi) [inline esteid/templates/esteid/test-new.html:27]: Info here why it can't be validated may be also beneficial for future us.

File: esteid/signing/views.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        """
+        return self.start_session(request, *args, **kwargs)
+
+    def patch(self, request, *args, **kwargs):
+        """
+        Handles session finish requests that require input data (ID Card)
+        """
+        return self.finish_session(request, *args, **kwargs)
+
+    def get(self, request, *args, **kwargs):
```
Comment (iharthi) [inline esteid/signing/views.py]: Arguably, `get` should be used for retrieving information and shouldn't cause side-effects on the server (like finishing the session). Browsers may use this in obscure cases when deciding if it is safe to re-play request for preview purposes, so I would still use patch for finalizing the session even without input - it shouldn't be any problem for front-end, and it makes it more explicit for developers and for browser that some state will be changed as result of this request.

File: esteid/signing/types.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    Interface for file objects that are added to signed containers
+
+    The container cares about 3 attributes:
+    * file name (base name only)
+    * mime type
+    * the content, for calculating hash.
+    """
+
+    def __init__(self, wraps: Union[str, File], mime_type: str = None, content: bytes = None, file_name: str = None):
```
Comment (iharthi) [inline esteid/signing/types.py:18]: :thinking: If we have explicit argument for `file_name`, maybe we should have also explicit argument for `file` instead of `wraps`, that can only be a `File`, and then we could allow either but not both to be empty. That would possibly make some code more readable/statically analyzable.

File: esteid/signing/signer.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+    The concrete Signer:
+    * does the necessary preparation of container/signature
+      for the files provided by the view;
+    * calls the associated services as necessary, to obtain the signature;
+    * finalizes the container.
+
+    The concrete Signer is also associated with an initial parameters type,
+    a session storage type, ... - akin to DRF serializers.
+
+    The only requirement to a concrete signer class is that it
```
Comment (iharthi) [inline esteid/signing/signer.py:28]: Typo. `is to` instead of `is that it`. Or `is that it inherits ... and registers ...`.

File: esteid/signing/README.md [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+        instance = self.get_object()
+        return [
+            DataFile("/path/to/document.doc", "application/ms-word-foo"),
+            DataFile(instance.document, "application/openoffice-file-bar")
+        ]
+
+    def save_container(self, container: Container, *args, **kwargs):
+        instance = self.get_object()
+        # Be sure to call `container_info(container)` before `container.finalize()`
+        instance.container_info = container_info(container)
```
Comment (iharthi) [inline esteid/signing/README.md:39]: A hint on what kind of field is required for `container_info` could be useful here.

File: esteid/signing/README.md [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+    - In: signed digest (**NOTE:** this requires using the PATCH method to perform the HTTP request)
+    - Do: finalize signature and container (OCSP, TSA)
+    - Out: (can be a redirect to the container download view)
+  * state between requests: the digest, temporary files
+* the external service flow with status: Mobile ID
+  * the `initial` request
+    - In: Personal ID code, Phone number
+    - Do: get user certificate from SK/MID; prepare the container; initiate the `sign` request to SK/MID
+    - Out: signature verification code
+  * the `status/final` request
```
Comment (iharthi) [inline esteid/signing/README.md:173]: I understand it is safe to replay this query until it changes response from `202` to `200`? Is it still safe to replay after it's `200` for second time? This kind of info here could be useful. (And I would probably still use patch always not get, but that is a discussion point, seems to be really an edge-case).

File: esteid/templates/esteid/test-new.html (hunk 6 above)
Comment (ykshatroff) [reply esteid/templates/esteid/test-new.html:27]: It is there in the readme. I guess a link to it would do, this view is there for testing not for documentation, I know assumption is mother of all evils but I still assume people won't end up on this page by accident.

File: esteid/signing/views.py (hunk 7 above)
Comment (ykshatroff) [reply esteid/signing/views.py]: Thanks, this makes total sense, overlooked the intent of GET. It makes it even easier. :)

File: esteid/signing/types.py (hunk 8 above)
Comment (ykshatroff) [reply esteid/signing/types.py:18]: The point of this wrapper was to make users type as little code as possible. 
I don't think there would be issues with static analyzers, and as for readability,
we could use a pure namedtuple/dataclass as earlier, but then it would require users to explicitly open and read files which surely makes their code less readable :) The `file_name` arg is there for the case where it can not be inferred from `wraps`, like `ContentFile`. I could document this to avoid questions

File: esteid/signing/signer.py (hunk 9 above)
Comment (ykshatroff) [reply esteid/signing/signer.py:28]: It's not a typo :) The correct English requires that there be `be` after `require that` :) 
https://english.stackexchange.com/questions/107384/require-that-every-lightbulb-have-has

File: esteid/signing/README.md (hunk 11 above)
Comment (ykshatroff) [reply esteid/signing/README.md:173]: Not sure about the meaning of _safe_ here, but as long as this endpoint forwards a status request to the upstream, I wouldn't call it safe. Also once a status has been received and container finalized, we clean up the session, any subsequent calls make no sense and will result in an error. But I'd rather keep it out of this part of the document as it's implementation details.

File: esteid/urls.py (hunk 4 above)
Comment (ykshatroff) [reply esteid/urls.py]: I thought this would help avoid accidentally including these urls in a project. But on a second thought, it might be actually useful to, so reverted.

File: esteid/templates/esteid/test-new.html (hunk 5 above)
Comment (ykshatroff) [reply esteid/templates/esteid/test-new.html:33]: Well I don't think this page should contain full documentation, after all this is just a note for testers as is the whole page.

**Review by Jyrno42 — APPROVED** [review]

**Review by iharthi — APPROVED** [review]
